core/src/main/scala/org/apache/spark/sql/delta/util/DeltaShufflePartitionsUtil.scala

```diff
@@ -0,0 +1,199 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.util
+
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.physical.{HashPartitioning, Partitioning, RoundRobinPartitioning}
+import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec
+import org.apache.spark.sql.execution.exchange.{ENSURE_REQUIREMENTS, ShuffleExchangeExec}
+import org.apache.spark.sql.execution.{CoalesceExec, PartialReducerPartitionSpec, SparkPlan}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.{MapOutputTrackerMaster, SparkEnv}
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Utility functions to rebalance shuffle partitions for OptimizeWrite.
+ */
+object DeltaShufflePartitionsUtil {
+
+  // scalastyle:off line.size.limit
+  /**
+   * Splits the skewed partition based on the map size and the target partition size
+   * after split, and create a list of `PartialMapperPartitionSpec`. Returns None if can't split.
+   *
+   * The function is copied from Spark 3.2:
+   *   https://github.com/apache/spark/blob/v3.2.1/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala#L376
+   * EDIT: Configurable smallPartitionFactor and mergedPartitionFactor.
+   *
+   * @shuffleId Shuffle ID for retrieve partition info.
+   * @reducerId Reducer ID (Partition ID) for retrieve partition info.
+   * @targetSize Target bin size for splitting.
+   * @smallPartitionFactor Threshold to avoid too small partition. If a partial partition is
+   *                       smaller than targetSize * smallPartitionFactor, merge it to an adjacent
+   *                       partition.
+   * @mergedPartitionFactor Threshold to avoid too large partition. If a partial partition is
+   *                        larger than targetSize * mergedPartitionFactor, do not merge other
+   *                        adjacent partitions to it.
+   */
+  // scalastyle:on
+  def createSkewPartitionSpecs(
+      shuffleId: Int,
+      reducerId: Int,
+      targetSize: Long,
+      smallPartitionFactor: Double,
+      mergedPartitionFactor: Double): Option[Seq[PartialReducerPartitionSpec]] = {
+    val mapPartitionSizes = getMapSizesForReduceId(shuffleId, reducerId)
+    if (mapPartitionSizes.exists(_ < 0)) return None
+    val mapStartIndices = splitSizeListByTargetSize(
+      mapPartitionSizes,
+      targetSize,
+      smallPartitionFactor,
+      mergedPartitionFactor)
+    if (mapStartIndices.length > 1) {
+      Some(mapStartIndices.indices.map { i =>
+        val startMapIndex = mapStartIndices(i)
+        val endMapIndex = if (i == mapStartIndices.length - 1) {
+          mapPartitionSizes.length
+        } else {
+          mapStartIndices(i + 1)
+        }
+        val dataSize = startMapIndex.until(endMapIndex).map(mapPartitionSizes(_)).sum
```

Comment:
Yes it's for both; make configurable and not rely on spark internal util which can change frequently. But I'm also okay to use Spark one.  I'm just waiting for databricks team to get back on this. (and auto compaction too)

Reply:
https://learn.microsoft.com/en-us/azure/synapse-analytics/spark/optimize-write-for-apache-spark

FYI, the code is being used in prod at least 6 months and no major issue so far.
We might need to increase binSize config / PARQUET COMPRESSION RATIO for larger files. (60~80mb avg for now)

Reply:
Hmm I tried building and using this myself and I don't seem to be getting my large partitions split, gonna add some more logging to try to see why/what's happening.

Reply:
Yes it's the case this approach (Spark rebalance logic) cannot handle. Because the unit of rebalancing is determined by the source partition layout. If the data is skewed or only few number of partitions, it cannot be rebalanced properly.
Please refer the figure in https://github.com/delta-io/delta/issues/1158#optimizewrite-partitioned-example

If the source dataframe consists of all same key=1 and 10GB of 1 partition, it cannot be split.
e.g. df.repartition(1, col("key")).select(col("key")).write.format("delta").parquet("path")

I removed redundant repartition execution plan on top of the child plan, but it cannot cover all the cases.

Reply:
The smallest unit of splitting is the single task map output for a single reducer ID right? That wasn't what I was seeing, where I had all my map tasks shuffle write < 1 GB, but I had some reducer tasks reading > 10 GB of shuffle data. After digging through how map output sizes work a little bit, I'm gonna try again and see if this is some weird side effect of `HighlyCompressedMapStatus` for large numbers of reducing partitions (> 2000 by default). Only thought is some weird effect of a lot of small blocks being "averaged out" to compute map output sizes per reducer.

Reply:
That appeared to be it. I dropped my shuffle partitions to 1k and it behaved as I would expect. Not sure how common my case would be with particular types of data skew, but maybe it would be good to log a warning if your shuffle partitions exceeds the threshold for using `HighlyCompressedMapStatus`, it can limit the ability to properly split skewed partitions, because average map output size is used for individual map reducer outputs, instead of the real value.